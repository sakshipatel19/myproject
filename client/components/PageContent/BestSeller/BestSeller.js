import React, { Component, Fragment } from "react";
import "./BestSeller.scss";
import DropDown from "./../../common/Dropdown/DropdownWithState";
import Card from "../../common/Card";
import Table from "../../common/Table";
import Tab from "../../common/Tab";
import LoadingIndicator from "../../common/LoadingSkeleton/LoadingSquare";
import DataLoadError from "../common/DataLoadError/DataLoadError";
import { withRouter } from "react-router";
import VerticalSeparator from "../../common/VerticalSeparator";
import { Link } from "react-router-dom";
import Icon from "../../common/Icon";
import { isNullOrUndefined } from "../../../utils/number";
import LoadingTable from "../../common/LoadingSkeleton/LoadingTable";

const defaultTableFilter = {
	sortBy: "rankNum",
	sortType: "ASC",
	pageNo: 1,
	pageSize: 15
};

class BestSeller extends Component {
	state = {
		tab: "",
		categoryNames: []
	};

	componentDidMount() {
		const tab = this.props.tabData || [];
		if (tab.length != 0) this.setState({ tab: tab[0].key });
	}
	static getDerivedStateFromProps(newProps, prevState) {
		if (
			newProps?.data?.data?.categoryNames !==
			prevState?.data?.data?.categoryNames &&
			newProps?.data?.data?.categoryNames?.length > 0
		)
			return { categoryNames: newProps?.data?.data?.categoryNames };
		return null;
	}
	getTableHeader = () => {
		return [
			{
				name: "Rank",
				key: "rankNum",
				isSortable: true,
				isNumber: false,
				style: { width: "8%" }
			},
			{
				name: "Brand",
				key: "brandName",
				isSortable: true,
				style: { width: "19%" }
			},
			{
				name: "Product Name",
				key: "productName",
				isSortable: false,
				iconName: "externalPage",
				style: { width: "41%" }
			},
			{
				name: "ASIN",
				key: "productId",
				isSortable: false,
				style: { width: "15%" }
			},
			{
				name: "Score",
				key: "score",
				isSortable: true,
				style: { width: "17%" }
			}
		];
	};
	DataRowComponent = row => {
		return (
			<tr>
				<td className={"table-cell-rank"}>{`#${row.rankNum}`}</td>
				<td className={"table-cell-brandName"}>{row.brandName}</td>
				<td
					className={"table-cell-productName"}
					onClick={() => window.open(`http://${row?.productUrl}`, "_blank")}
				>
					{row.productName}
				</td>
				<td className={"table-cell-productId"}>{row.productId}</td>
				<td className={"table-cell-score"}>{this.getScore(row.score)}</td>
			</tr>
		);
	};

	getScore = score => {
		return isNullOrUndefined(score) ? "NA" : `${score}%`;
	};

	onTabChange = tab => {
		this.setState({ tab });
		this.props.getBestSellersListOnTabChange(
			this.props.data?.data?.selectedCategory,
			tab
		);
	};

	handleCategorySelection = selectedCategory => {
		this.props.getBestSellersList(selectedCategory, this.state.tab);
	};

	createUniqueKeyString = obj => `${obj.productId}__${obj.rank}`;

	onClickBestSellerTable = () => {
		this.props.onClickBestSellerTable(
			this.props.data?.data?.selectedCategory,
			this.state.categoryNames
		);
	};
	render() {
		const {
			onRetryClick,
			showDetailPageLink,
			tableHeader,
			DataRowComponent,
			createUniqueKeyString,
			tableFilter,
			enable
		} = this.props;
		const { data } = this.props?.data || {};
		let bestSellerBrandDataList = data?.bestSellerBrandDataList || [];

		const tabData = this.props.tabData || [];
		const bestseller = this.props?.data || {};
		return (enable ?
			<div className="bestseller-container section-header-main">
				<Card>
					<div
						className={`${"card-header"} ${tabData.length != 0 ? "tabPresent" : ""
							}`}
					>
						<div className="card-title section-header-title">
							{"Top Sellers"}
						</div>
						<div className="bestseller-tab-container">
							{tabData.length != 0 && (
								<div className="bestseller-tab">
									<Tab
										data={tabData}
										selectedTab={this.state.tab}
										onTabChange={this.onTabChange}
									/>
								</div>
							)}
							<div className={"bestseller-dropdown-container"}>
								<DropDown
									options={data?.categoryNames?.map(value => ({
										value,
										label: value
									}))}
									onSelect={this.handleCategorySelection}
									selectedOption={data?.selectedCategory}
								/>
								{showDetailPageLink && (
									<Fragment>
										<VerticalSeparator
											className={"best-seller-details-separator"}
										/>
										<Link
											to={`${this.props.location.pathname}/bestseller-details`}
											className={"best-seller-details-link"}
											onClick={this.onClickBestSellerTable}
										>
											TOP SELLER KEYWORDS
											<Icon
												name={"sortRight"}
												size={12}
												iconClass={"best-sller-detail-icon"}
											/>
											<Icon
												name={"sortRightBlue"}
												size={12}
												iconClass={"best-sller-detail-icon-hover"}
											/>
										</Link>
									</Fragment>
								)}
							</div>
						</div>
					</div>
					{bestseller.fetching ? (
						<LoadingTable className="bestseller-loading-indicator" />
					) : bestseller.error ? (
						<DataLoadError handleRetry={onRetryClick} />
					) : (
								<div className="bestseller-detailpage-container">
									<Table
										header={tableHeader || this.getTableHeader()}
										data={bestSellerBrandDataList}
										tableFilter={tableFilter || defaultTableFilter}
										DataRowComponent={DataRowComponent || this.DataRowComponent}
										createUniqueKeyString={
											createUniqueKeyString || this.createUniqueKeyString
										}
										isInternalSorting
									/>
								</div>
							)}
				</Card>
			</div> : null
		);
	}
}

export default withRouter(BestSeller);
