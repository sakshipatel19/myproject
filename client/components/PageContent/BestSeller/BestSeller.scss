@import "../../../styles/colors.scss";

.bestseller-container {
	margin-top: 50px;
	margin-bottom: 50px;
	.card-container {
		.card-header.tabPresent {
			flex-direction: column;
			.bestseller-tab-container {
				display: flex;
				margin-top: 43px;
				width: 100%;
				.bestseller-dropdown-container {
					margin-left: auto;
					margin-bottom: 4px;
					padding-right: 0;
				}
			}
		}
		.bestseller-dropdown-container {
			display: flex;
			flex-direction: row;
			justify-content: flex-end;
			padding-right: 30px;
			align-items: center;
			align-content: center;
			margin-top: 0px;
			.dropdown-container {
				width: 160px;
			}
			.best-seller-details-separator {
				margin: 0 40px;
				height: 16px;
				background-color: $verticalSeparatorBGColor;
			}
			.best-seller-details-link {
				text-decoration: none;
				color: $primaryBlackColor;
				font-family: "NotoSans-SemiBold";
				font-size: 12px;
				letter-spacing: 2px;
				.best-sller-detail-icon-hover {
					margin-left: 10px;
					vertical-align: middle;
					display: none;
				}
				.best-sller-detail-icon {
					margin-left: 10px;
					vertical-align: middle;
				}
				&:hover {
					color: $linkHoveredColor;
					.best-sller-detail-icon-hover {
						display: inline;
					}
					.best-sller-detail-icon {
						display: none;
					}
				}
			}
		}
		.bestseller-detailpage-container {
			margin-top: 35px;
			margin-bottom: 10px;

			.table-cell-productName {
				width: 48.94%;
				padding-right: 8.77%;
			}
			.table-cell-productName:hover {
				cursor: pointer;
				color: $tableTextHover;
			}
			thead {
				.table-sort-image {
					padding-left: 10px;
				}
				th:last-child {
					padding-left: 9px;
					padding-right: 30px;
					text-align: right;
				}
			}
			td:last-child {
				padding-right: 30px;
				text-align: right;
			}
		}
		.bestseller-loading-indicator {
			min-height: 450px;
		}
	}
}

.bestseller-keyword-detail-container {
	width: 100%;
	.bestseller-keyword-search-container {
		display: flex;
		.bestseller-keyword-search {
			width: 70%;
			margin-right: 3%;
		}
		.bestseller-keyword-dropdown {
			.dropdown-container {
				width: 137px;
			}
		}
		.bestseller-keyword-filter {
			display: flex;
			width: 10%;
			height: 18px;
			margin: 6px 0 6px auto;
			padding-left: 3%;
			border-left: 1px solid $tabContainerBorderBottom;
			color: $primaryBlackColor;
			font-family: "NotoSans-SemiBold";
			font-size: 12px;
			font-weight: 600;
			letter-spacing: 1px;
			line-height: 18px;
			margin-bottom: 5px;
			cursor: pointer;
			.icon-class {
				margin-left: 3%;
				margin-right: 10%;
			}
		}
	}
	.bestseller-keyword-table-container {
		.common-table-container {
			.aligncenter {
				height: 62px;
			}
			thead {
				th:nth-of-type(6),
				th:nth-of-type(7),
				th:nth-of-type(8) {
					text-align: left;
					padding-right: 0;
					padding-left: 37px;
				}
				th:nth-of-type(4) {
					text-align: left;
					padding-right: 16px;
				}

				th:nth-of-type(5),
				th:nth-of-type(6),
				th:nth-of-type(7) {
					color: $tableHeadingLightColor;
				}
			}
			tbody {
				td {
					text-align: left;
				}
				.first-column {
					padding-right: 3%;
				}
				.last-column {
					padding-right: 10px;
				}
				.number-text {
					text-align: right;
				}
			}
		}
	}
}
.bestseller-by-search {
	.bestseller-container {
		.card-container {
			.bestseller-detailpage-container {
				thead {
					.table-sort-image {
						position: absolute;
						top: 15px;
					}
				}
				.common-table-container {
					.number-text {
						&.sortSlelected {
							position: relative;
						}
					}
				}
			}
		}
	}
}
