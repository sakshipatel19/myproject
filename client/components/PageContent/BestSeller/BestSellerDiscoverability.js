import React from "react";
import { isNullOrUndefined } from "../../../utils/number";
import BestSeller from "./BestSeller";

const BestSellerDisc = props => {
	const tableFilter = {
		sortBy: "overallScore",
		sortType: "DESC",
		pageNo: 1,
		pageSize: 15
	};

	const DataRowComponent = row => {
		return (
			<tr>
				{/* <td>{`#${row.rank}`}</td> */}
				<td>{row.brand}</td>
				<td className={"number-text"}>{getScore(row.overallScore)}</td>
				<td className={"number-text"}>{getScore(row.organicScore)}</td>
				<td className={"number-text"}>{getScore(row.paidScore)}</td>
				<td></td>
			</tr>
		);
	};

	const getScore = score => {
		return isNullOrUndefined(score) ? "NA" : `${score}%`;
	};

	const createUniqueKeyString = row => `${row.rank}_${row.brand}`;


	return (props.enable ?
		<div className="bestseller-by-search">
			<BestSeller
				{...props}
				showDetailPageLink
				DataRowComponent={DataRowComponent}
				createUniqueKeyString={createUniqueKeyString}
				tableFilter={tableFilter}
			/>
		</div> : null
	);
};

export default BestSellerDisc;
