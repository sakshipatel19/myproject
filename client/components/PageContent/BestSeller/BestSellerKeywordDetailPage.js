import React, { Component } from "react";
import PageContainer from "../PageContainer";
import { BESTSELLER_DETAILS, pageTitle } from "../../../constants/PageHeaders";
import Card from "../../common/Card";
import Search from "../../common/Search/Search";
import Table from "../../common/Table";
import Dropdown from "../../common/Dropdown/DropdownWithState";
import Icon from "../../common/Icon";
import LoadingTable from "../../common/LoadingSkeleton/LoadingTable";
import DataLoadError from "../common/DataLoadError/DataLoadError";
import NoResultsMessage from "../common/NoResultsMessage";
import "./BestSeller.scss";

const defaultTableFilter = {
	pageNo: 1,
	pageSize: 15,
	sortBy: "keyword",
	sortType: "ASC"
};
class BestSellerKeywordDetail extends Component {
	state = {
		tableFilter: defaultTableFilter
	};
	getCommonPayload = (selectedCategory = "") => {
		return {
			tableFilter: this.state.tableFilter,
			selectedCategory: selectedCategory
		};
	};
	componentDidMount() {
		let payload = {
			tableFilter: this.state.tableFilter,
			selectedCategory: this.props.selectedCategory,
			categoryNames: this.props.categoryNames
		};
		this.props.setSelectedPageInConfig(BESTSELLER_DETAILS);
		this.props.getBestSellerKeywordDetail(payload);
		window.scrollTo(0, 0);
	}
	onSeachKeywordsChange = searchList => {
		let tableFilter = {
			...defaultTableFilter,
			searchTerms: searchList
		};
		this.onTableFilterChange(tableFilter);
	};

	onTableFilterChange = tableFilter => {
		const payload = {
			tableFilter,
			selectedCategory: this.props.data?.data?.selectedCategory
		};
		this.setState({ tableFilter });
		this.props.getBestSellerKeywordDetail(payload);
	};
	onCategoryDropdownChange = selectedCategory => {
		this.setState({ tableFilter: defaultTableFilter });
		this.props.getBestSellerKeywordDetail(
			this.getCommonPayload(selectedCategory)
		);
	};

	createUniqueKeyString = obj =>
		`${obj.keyword}_${obj.keywordType}_${obj.brand}`;

	render() {
		const {
			data,
			tableHeader,
			DataRowComponent,
			getBestSellerKeywordDetail,
			categoryNames,
			enable
		} = this.props;
		const keywordsData = data?.data || {};
		const paginator = keywordsData?.paginator || {};
		const shareOfShelfList = keywordsData?.shareOfShelfList || [];
		const searchTerms = this.state.tableFilter?.searchTerms || [];
		const selectedCategory = keywordsData?.selectedCategory || "";

		return (enable ?
			<PageContainer title={pageTitle[BESTSELLER_DETAILS]}>
				<div className="bestseller-keyword-detail-container">
					<Card>
						<div className="bestseller-keyword-search-container">
							<div className="bestseller-keyword-search">
								<Search
									placeholder={"Type to search here..."}
									getSearchedKeywordList={this.onSeachKeywordsChange}
									searchKeywordList={searchTerms || []}
								/>
							</div>
							<div className="bestseller-keyword-dropdown">
								<Dropdown
									options={categoryNames?.map(value => ({
										value,
										label: value
									}))}
									onSelect={this.onCategoryDropdownChange}
									selectedOption={selectedCategory}
								/>
							</div>
							<div className="bestseller-keyword-filter">
								<Icon name={"funnel"} size={16} />
								FILTER
							</div>
						</div>
						{data?.fetching ? (
							<LoadingTable />
						) : data?.error ? (
							<DataLoadError
								handleRetry={() =>
									getBestSellerKeywordDetail(
										this.getCommonPayload(selectedCategory)
									)
								}
							/>
						) : shareOfShelfList?.length === 0 ? (
							<NoResultsMessage />
						) : (
										<div className="bestseller-keyword-table-container">
											<Table
												header={tableHeader}
												data={shareOfShelfList}
												tableFilter={this.state.tableFilter}
												onTableFilterChange={this.onTableFilterChange}
												DataRowComponent={DataRowComponent}
												paginator={paginator}
												ispaginationRequired={paginator?.totalPages > 1}
												createUniqueKeyString={this.createUniqueKeyString}
											/>
										</div>
									)}
					</Card>
				</div>
			</PageContainer> : null
		);
	}
}

export default BestSellerKeywordDetail;
