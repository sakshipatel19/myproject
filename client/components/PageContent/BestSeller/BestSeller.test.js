import React from "react";
import { shallow } from "enzyme";

import BestSeller from "./BestSeller";

const bestSellerBrandDataListEntry = {
	brandName: "Tylenol",
	productId: "ABCD",
	productName: "abcd efgh ijkl",
	productUrl: "www.amazon.com",
	rank: "#1",
	score: 75
};
const bestSeller = {
	fetching: null,
	error: null,
	data: {
		bestSellerBrandDataList: [bestSellerBrandDataListEntry],
		categoryNames: ["Calcium", "Antacids"]
	}
};
const tabData = [
	{
		key: "price",
		label: "PRICE"
	},
	{
		key: "promotion",
		label: "PROMOTION"
	}
];
const baseProps = {
	data: bestSeller,
	tabData: tabData,
	getBestSellersListOnTabChange: jest.fn(),
	onRetryClick: jest.fn(),
	getBestSellersList: jest.fn()
};
describe("BestSeller tests", () => {
	it("BestSeller should render correctly", () => {
		const wrapper = shallow(<BestSeller {...baseProps} />);
        expect(wrapper).toMatchSnapshot();
        //cover all the functions
		wrapper.instance().handleCategorySelection();
		wrapper.instance().onTabChange();
        wrapper.instance().DataRowComponent(bestSellerBrandDataListEntry);
		expect(baseProps.getBestSellersList).toHaveBeenCalledTimes(1);
        expect(baseProps.getBestSellersListOnTabChange).toHaveBeenCalledTimes(1);
        expect(baseProps.DataRowComponent).toHaveBeenCalledTimes(1);
	});
	it("BestSeller should render correctly when no props are received", () => {
		const wrapper = shallow(<BestSeller />);
		expect(wrapper).toMatchSnapshot();
	});
});
