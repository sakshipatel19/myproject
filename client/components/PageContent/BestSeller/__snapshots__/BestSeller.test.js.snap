// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`BestSeller tests BestSeller should render correctly 1`] = `
<div
  className="bestseller-container section-header-main"
>
  <Card>
    <div
      className="card-header tabPresent"
    >
      <div
        className="card-title section-header-title"
      >
        Top Sellers
      </div>
      <div
        className="bestseller-tab-container"
      >
        <div
          className="bestseller-tab"
        >
          <Tab
            data={
              Array [
                Object {
                  "key": "price",
                  "label": "PRICE",
                },
                Object {
                  "key": "promotion",
                  "label": "PROMOTION",
                },
              ]
            }
            onTabChange={[Function]}
            selectedTab="price"
          />
        </div>
        <div
          className="bestseller-dropdown-container"
        >
          <DropdownWithState
            onSelect={[Function]}
            options={
              Array [
                Object {
                  "label": "Calcium",
                  "value": "Calcium",
                },
                Object {
                  "label": "Antacids",
                  "value": "Antacids",
                },
              ]
            }
          />
        </div>
      </div>
    </div>
    <div
      className="bestseller-detailpage-container"
    >
      <Table
        DataRowComponent={[Function]}
        data={
          Array [
            Object {
              "brandName": "Tylenol",
              "productId": "ABCD",
              "productName": "abcd efgh ijkl",
              "productUrl": "www.amazon.com",
              "rank": "#1",
              "score": 75,
            },
          ]
        }
        header={
          Array [
            Object {
              "isNumber": false,
              "isSortable": true,
              "key": "rank",
              "name": "Rank",
              "style": Object {
                "width": "10.57%",
              },
            },
            Object {
              "isSortable": true,
              "key": "brandName",
              "name": "Brand",
              "style": Object {
                "width": "22.6%",
              },
            },
            Object {
              "iconName": "externalPage",
              "isSortable": false,
              "key": "productName",
              "name": "Product Name",
              "style": Object {
                "paddingRight": "8.77%",
                "width": "48.94%",
              },
            },
            Object {
              "isSortable": false,
              "key": "productId",
              "name": "ASIN",
              "style": Object {
                "width": "16.8%",
              },
            },
            Object {
              "isSortable": true,
              "key": "score",
              "name": "Score",
              "style": Object {
                "width": "4%",
              },
            },
          ]
        }
        isInternalSorting={true}
        tableFilter={
          Object {
            "pageNo": 1,
            "pageSize": 15,
            "sortBy": "rank",
            "sortType": "ASC",
          }
        }
      />
    </div>
  </Card>
</div>
`;

exports[`BestSeller tests BestSeller should render correctly when no props are received 1`] = `
<div
  className="bestseller-container section-header-main"
>
  <Card>
    <div
      className="card-header "
    >
      <div
        className="card-title section-header-title"
      >
        Top Sellers
      </div>
      <div
        className="bestseller-tab-container"
      >
        <div
          className="bestseller-dropdown-container"
        >
          <DropdownWithState
            onSelect={[Function]}
          />
        </div>
      </div>
    </div>
    <div
      className="bestseller-detailpage-container"
    >
      <Table
        DataRowComponent={[Function]}
        data={Array []}
        header={
          Array [
            Object {
              "isNumber": false,
              "isSortable": true,
              "key": "rank",
              "name": "Rank",
              "style": Object {
                "width": "10.57%",
              },
            },
            Object {
              "isSortable": true,
              "key": "brandName",
              "name": "Brand",
              "style": Object {
                "width": "22.6%",
              },
            },
            Object {
              "iconName": "externalPage",
              "isSortable": false,
              "key": "productName",
              "name": "Product Name",
              "style": Object {
                "paddingRight": "8.77%",
                "width": "48.94%",
              },
            },
            Object {
              "isSortable": false,
              "key": "productId",
              "name": "ASIN",
              "style": Object {
                "width": "16.8%",
              },
            },
            Object {
              "isSortable": true,
              "key": "score",
              "name": "Score",
              "style": Object {
                "width": "4%",
              },
            },
          ]
        }
        isInternalSorting={true}
        tableFilter={
          Object {
            "pageNo": 1,
            "pageSize": 15,
            "sortBy": "rank",
            "sortType": "ASC",
          }
        }
      />
    </div>
  </Card>
</div>
`;
